src/lib/storage/buffer/volatile_region.cpp

```diff
@@ -0,0 +1,238 @@
+}
+
+PageSizeType VolatileRegion::size_type() const {
+  return _size_type;
+}
+
+void VolatileRegion::reuse(const PageID page_id) {
+  _unprotect_page(page_id);
+  // On OS X, we can use MADV_FREE_REUSE to update the memory accounting.
+  // Source: https://bugs.chromium.org/p/chromium/issues/detail?id=823915
+#ifdef __APPLE__
+  const auto ptr = get_page(page_id);
+  if (madvise(ptr, page_id.byte_count(), MADV_FREE_REUSE) < 0) {
+    const auto error = errno;
+    Fail("Failed to call madvise(MADV_FREE_REUSE): " + strerror(error));
+  }
+#endif
+}
+
+void VolatileRegion::_protect_page(const PageID page_id) {
+  if constexpr (ENABLE_MPROTECT) {
+    DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+    auto data = get_page(page_id);
+    if (mprotect(data, page_id.byte_count(), PROT_NONE) != 0) {
+      const auto error = errno;
+      Fail("Failed to mprotect: " + strerror(error));
+    }
+  }
+}
+
+void VolatileRegion::_unprotect_page(const PageID page_id) {
+  if constexpr (ENABLE_MPROTECT) {
+    DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+    auto data = get_page(page_id);
+    if (mprotect(data, page_id.byte_count(), PROT_READ | PROT_WRITE) != 0) {
+      const auto error = errno;
+      Fail("Failed to mprotect: " + strerror(error));
+    }
+  }
+}
+
+inline std::size_t get_os_page_size() {
+  return std::size_t(sysconf(_SC_PAGESIZE));
+}
+
+VolatileRegion::MappedRegion VolatileRegion::create_mapped_region() {
+  Assert(bytes_for_size_type(MIN_PAGE_SIZE_TYPE) >= get_os_page_size(),
+         "Smallest page size does not fit into an OS page: " + std::to_string(get_os_page_size()));
+#ifdef __APPLE__
+  const int flags = MAP_PRIVATE | MAP_ANON | MAP_NORESERVE;
+#elif __linux__
+  const int flags = MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE;
+#endif
+  const auto mapped_memory = static_cast<std::byte*>(
+      mmap(NULL, VolatileRegion::DEFAULT_RESERVED_VIRTUAL_MEMORY, PROT_READ | PROT_WRITE, flags, -1, 0));
+
+  if (mapped_memory == MAP_FAILED) {
+    const auto error = errno;
+    Fail("Failed to map volatile pool region: " + strerror(error));
+  }
+
+  return MappedRegion(mapped_memory, DEFAULT_RESERVED_VIRTUAL_MEMORY);
+}
+
+std::array<std::shared_ptr<VolatileRegion>, PAGE_SIZE_TYPES_COUNT> VolatileRegion::create_volatile_regions(
+    VolatileRegion::MappedRegion mapped_region) {
+  auto array = std::array<std::shared_ptr<VolatileRegion>, PAGE_SIZE_TYPES_COUNT>{};
+
+  // Ensure that every region has the same amount of virtual memory
+  // Round to the next multiple of the largest page size
+  for (auto i = size_t{0}; i < PAGE_SIZE_TYPES_COUNT; i++) {
+    auto region = MappedPageRegion{mapped_region.data() + DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION * i,
+                                   DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION};
+    array[i] = std::make_shared<VolatileRegion>(magic_enum::enum_value<PageSizeType>(i), region);
```

Comment:
I haven't seen the usage yet, but I am surprised to see shared over unique pointers here.